--- a/arch/arm64/boot/dts/qcom/ipq8074.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq8074.dtsi
@@ -106,7 +106,6 @@
 			reg = <0x0 0x4a400000 0x0 0x00200000>;
 		};
 
-
 		uboot@4a600000 {
 			no-map;
 			reg = <0x0 0x4a600000 0x0 0x00400000>;
@@ -129,17 +128,17 @@
 
 		q6_region: wcnss@4b000000 {
 			no-map;
-			reg = <0x0 0x4b000000 0x0 0x05f00000>;
+			reg = <0x0 0x4b000000 0x0 0x03700000>;
 		};
 
-		q6_etr_region: q6_etr_dump@50f00000 {
+		q6_etr_region: q6_etr_dump@4e700000 {
 			no-map;
-			reg = <0x0 0x50f00000 0x0 0x00100000>;
+			reg = <0x0 0x4e700000 0x0 0x00100000>;
 		};
 
-		m3_dump@51000000 {
+		m3_dump@4e800000 {
 			no-map;
-			reg = <0x0 0x51000000 0x0 0x100000>;
+			reg = <0x0 0x4e800000 0x0 0x00100000>;
 		};
 
 	};
