--- a/drivers/net/phy/realtek/realtek_main.c
+++ b/drivers/net/phy/realtek/realtek_main.c
@@ -218,6 +218,7 @@
 #define RTL_8264B				0x001cc813
 
 #define RTL8261C_CE_MODEL		0x00
+#define RTL8261D_MODEL			0x81
 #define RTL8261X_IMR			0xa4d2
 #define RTL8261X_ISR			0xa4d4
 #define RTL8261X_INT_AUTONEG_ERROR	BIT(0)
@@ -384,6 +385,11 @@ static int rtl8261x_probe(struct phy_dev
 		phydev_info(phydev, "RTL8261C detected (sub_id 0x%02x)\n", sub_phy_id);
 		break;
 
+	case RTL8261D_MODEL:
+		priv->fw_name = RTL8261C_CE_FW_NAME;
+		phydev_info(phydev, "RTL8261D detected (sub_id 0x%02x)\n", sub_phy_id);
+		break;
+
 	default:
 		phydev_err(phydev, "Unknown sub_id 0x%02x\n", sub_phy_id);
 		return -ENODEV;
